k9mail/src/main/java/com/fsck/k9/mailstore/MessageViewInfoExtractor.java

```diff
@@ -71,43 +73,58 @@ public static MessageViewInfoExtractor getInstance() {
+
+        if (cryptoContentPart == null) {
             if (annotations != null && !annotations.isEmpty()) {
-                Timber.e("Got message annotations but no crypto root part!");
+                Timber.e("Got crypto message annotations but no crypto root part!");
             }
-            rootPart = message;
-            cryptoResultAnnotation = null;
-            extraParts = null;
+            return extractSimpleMessageForView(message, message);
         }
 
-        List<AttachmentViewInfo> attachmentInfos = new ArrayList<>();
-        ViewableExtractedText viewable = extractViewableAndAttachments(
-                Collections.singletonList(rootPart), attachmentInfos);
+        boolean isOpenPgpEncrypted =
+                MimeUtility.isSameMimeType(cryptoContentPart.getMimeType(), "multipart/encrypted") &&
+                MimeUtility.isSameMimeType(getHeaderParameter(cryptoContentPart.getContentType(), "protocol"),
+                        "application/pgp-encrypted");
+        if (!K9.isOpenPgpProviderConfigured() && isOpenPgpEncrypted) {
+            CryptoResultAnnotation noProviderAnnotation = CryptoResultAnnotation.createErrorAnnotation(
+                    CryptoError.OPENPGP_ENCRYPTED_NO_PROVIDER, null);
+            return MessageViewInfo.createWithErrorState(message, false).withCryptoData(noProviderAnnotation, null, null);
+        }
 
-        List<AttachmentViewInfo> extraAttachmentInfos = new ArrayList<>();
-        String extraViewableText = null;
-        if (extraParts != null) {
-            ViewableExtractedText extraViewable =
-                    extractViewableAndAttachments(extraParts, extraAttachmentInfos);
-            extraViewableText = extraViewable.text;
+        CryptoResultAnnotation cryptoContentPartAnnotation = annotations != null ? annotations.get(cryptoContentPart) : null;
+        if (cryptoContentPartAnnotation != null) {
+            return extractCryptoMessageForView(message, extraParts, cryptoContentPart, cryptoContentPartAnnotation);
+        }
+
+        return extractSimpleMessageForView(message, message);
+    }
+
+    private MessageViewInfo extractCryptoMessageForView(Message message,
+            ArrayList<Part> extraParts, Part cryptoContentPart, CryptoResultAnnotation cryptoContentPartAnnotation)
+            throws MessagingException {
+        if (cryptoContentPartAnnotation != null && cryptoContentPartAnnotation.hasReplacementData()) {
+            cryptoContentPart = cryptoContentPartAnnotation.getReplacementData();
         }
 
-        AttachmentResolver attachmentResolver = AttachmentResolver.createFromPart(rootPart);
+        List<AttachmentViewInfo> extraAttachmentInfos = new ArrayList<>();
+        ViewableExtractedText extraViewable = extractViewableAndAttachments(extraParts, extraAttachmentInfos);
+
+        MessageViewInfo messageViewInfo = extractSimpleMessageForView(message, cryptoContentPart);
```

Comment:
cryptoContentPart is 'decryptoedContentPart' ?

Reply:
well, sometimes. the content part that is actually displayed may be replaced with something else from the crypto annotation (four lines above)

Reply:
Hence decyptoed rather than decrypted. Too subtle I guess. Maybe `viewableVersionOfContentPart`

As in:

```
Part viewableVersionOfContentPart;
if (...)
   viewableVersionOfContentPart = cryptoContentPartAnnotation.getReplacementData();
else
  viewableVersionOfContentPart = cryptoContentPart;
```

Reply:
Eh. It's a 9 line method, I don't think we have to be too picky here